source/opt/decoration_manager.cpp

```diff
@@ -18,14 +18,91 @@
+  auto context = module_->context();
+  std::vector<ir::Instruction*> insts_to_kill;
+  const bool is_group = !decorations_info.decorate_insts.empty();
+
+  // Schedule all direct decorations for removal if instructed as such by
+  // |pred|.
+  for (ir::Instruction* inst : decorations_info.direct_decorations)
+    if (pred(*inst)) insts_to_kill.push_back(inst);
+
+  // For all groups being directly applied to |id|, remove |id| (and the
+  // literal if |inst| is an OpGroupMemberDecorate) from the instruction
+  // applying the group.
+  std::unordered_set<const ir::Instruction*> indirect_decorations_to_remove;
+  for (ir::Instruction* inst : decorations_info.indirect_decorations) {
+    assert(inst->opcode() == SpvOpGroupDecorate ||
+           inst->opcode() == SpvOpGroupMemberDecorate);
+    const uint32_t stride = inst->opcode() == SpvOpGroupDecorate ? 1u : 2u;
+    bool was_modified = false;
+    for (uint32_t i = 1u; i < inst->NumInOperands();) {
+      if (inst->GetSingleWordInOperand(i) != id) {
+        i += stride;
+        continue;
+      }
+
+      const uint32_t last_operand_index = inst->NumInOperands() - stride;
+      if (i < last_operand_index)
+        inst->GetInOperand(i) = inst->GetInOperand(last_operand_index);
+      // Remove the associated literal, if it exists.
+      if (stride == 2u) {
+        if (i < last_operand_index)
+          inst->GetInOperand(i + 1u) =
+              inst->GetInOperand(last_operand_index + 1u);
+        inst->RemoveInOperand(last_operand_index + 1u);
+      }
+      inst->RemoveInOperand(last_operand_index);
+      was_modified = true;
+    }
+
+    // If the instruction has no targets left, remove the instruction
+    // altogether.
+    if (inst->NumInOperands() == 1u) {
+      indirect_decorations_to_remove.emplace(inst);
+      insts_to_kill.push_back(inst);
+    } else if (was_modified) {
+      indirect_decorations_to_remove.emplace(inst);
+      context->AnalyzeUses(inst);
+    }
+  }
+
+  // Schedule all instructions applying the group for removal if instructed as
+  // such by |pred|.
+  for (ir::Instruction* inst : decorations_info.decorate_insts)
+    insts_to_kill.push_back(inst);
+
+  auto& indirect_decorations = decorations_info.indirect_decorations;
+  indirect_decorations.erase(
+      std::remove_if(
+          indirect_decorations.begin(), indirect_decorations.end(),
+          [&indirect_decorations_to_remove](const ir::Instruction* inst) {
+            return indirect_decorations_to_remove.count(inst);
+          }),
+      indirect_decorations.end());
+
+  for (ir::Instruction* inst : insts_to_kill) context->KillInst(inst);
+
+  if (decorations_info.direct_decorations.empty() &&
+      decorations_info.indirect_decorations.empty() &&
+      decorations_info.decorate_insts.empty()) {
+    id_to_decoration_insts_.erase(ids_iter);
+  }
+
+  // Remove the OpDecorationGroup defining this group.
+  if (is_group) context->KillInst(context->get_def_use_mgr()->GetDef(id));
```

Comment:
You cannot delete this unless all uses of `id` are also killed, but what if `pred` returned false for something?

Reply:
This should be in the if statement right above it, to guard against that eventuality.